### PR TITLE
[PM-11170] Inline menu should not show incomplete login item along passkey

### PR DESCRIPTION
**Pull request by cagonzalezcs**
## 🎟️ Tracking

https://bitwarden.atlassian.net/browse/PM-11170

## 📔 Objective

This work addresses some issues found with authenticating using passkeys from the inline menu. It ensures that login ciphers are not displayed alongside passkeys if the login cipher does not contain a password. It also ensures that we do not display passkeys on fields that do not contain a `webauth` autocomplete value.

Along with the above, some changes were made to how we handle FIDO2 assertion requests through a conditionally mediated UI. As we have rolled back user verification for passkeys, we are wanting to do the same with the inline menu implementation. With these changes, the only time the fido2 authenticator will open with a conditional mediated authentication is when the user has `master password reprompt` set for the cipher. 

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts [hunk 1]
```diff
@@ -242,7 +242,15 @@ export class Fido2AuthenticatorService implements Fido2AuthenticatorServiceAbstr
         throw new Fido2AuthenticatorError(Fido2AuthenticatorErrorCode.NotAllowed);
       }
 
-      let response = { cipherId: cipherOptions[0].id, userVerified: false };
+      // NOTE: For now, we are defaulting to a userVerified status of `true`. This will allow for mediated conditional
```
Comment (cagonzalezcs) [inline libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:245]: @coroiu 

Definitely let me know if this doesn't make sense, or should be implemented differently.

In effect, we're wanting to ensure that conditional mediated requests do not trigger the authenticator unless the cipher is set to have a master password reprompt.

**Review by jprusik — APPROVED** [review]

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts [hunk 2]
```diff
@@ -242,7 +242,16 @@ export class Fido2AuthenticatorService implements Fido2AuthenticatorServiceAbstr
...
+      // NOTE: For now, we are defaulting to a userVerified status of `true` when the request
+      // is for a conditionally mediated authentication. This will allow for mediated conditional
+      // authentication to function without requiring user interaction. This is a product
+      // decision, rather than a decision based on the expected technical specifications. This
+      // will be updated to have a conditional of `params.requireUserVerification || !params.assumeUserPresence`
+      // once the notification bar rework check is implemented.
+      let response = {
+        cipherId: cipherOptions[0].id,
+        userVerified: params.conditionalMediatedAuth,
+      };
```
Comment (coroiu) [inline libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:254]: issue: I'm not a big fan of changing the fido2 services, since the more these diverge from the spec, the harder it will become to switch to the SDK.

suggestion: Move the faking to `BrowserFido2UserInterfaceService.pickCredential` https://github.com/bitwarden/clients/blob/e60669d4bb574cee028c44d808da6e87d43d8b2b/apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts#L241-L257

You will have to add a new parameter which informs `pickCredential` if this is a mediated request, in which case it should just short circuit without opening any windows.

I'm not sure how much prettier this is, but it'll be more consistent with how this will have to be solved in the SDK.

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts [hunk 3]
```diff
@@ -242,7 +242,16 @@ export class Fido2AuthenticatorService implements Fido2AuthenticatorServiceAbstr
...
-      let response = { cipherId: cipherOptions[0].id, userVerified: false };
+      // NOTE: For now, we are defaulting to a userVerified status of `true` when the request
+      // is for a conditionally mediated authentication. This will allow for mediated conditional
+      // authentication to function without requiring user interaction. This is a product
+      // decision, rather than a decision based on the expected technical specifications. This
+      // will be updated to have a conditional of `params.requireUserVerification || !params.assumeUserPresence`
+      // once the notification bar rework check is implemented.
+      let response = {
+        cipherId: cipherOptions[0].id,
+        userVerified: params.conditionalMediatedAuth,
```
Comment (coroiu) [inline libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:253]: issue: IF we don't move the faking to `BrowserFido2UserInterfaceService.pickCredential` then we need to fix this row. From what I can see this will always return `userVerified = true` for conditional requests. We shouldn't be lying like this if we don't have to, i.e. only return true if `params.requireUserVerification && params.conditionalMediatedAuth`

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts [hunk 4]
```diff
@@ -329,9 +338,12 @@ export class Fido2AuthenticatorService implements Fido2AuthenticatorServiceAbstr
...
     cipherOptions: CipherView[],
   ): boolean {
+    const requiresUserVerification =
+      !params.conditionalMediatedAuth &&
+      (params.requireUserVerification || !params.assumeUserPresence);
+
     return (
-      params.requireUserVerification ||
-      !params.assumeUserPresence ||
+      requiresUserVerification ||
```
Comment (coroiu) [inline libs/common/src/platform/services/fido2/fido2-authenticator.service.ts:346]: if we move the faking to the browser user interface service then we can revert this change

File: libs/common/src/platform/services/fido2/fido2-authenticator.service.ts (hunk 2 above)
Comment (cagonzalezcs) [reply libs/common/src/platform/services/fido2/fido2-authenticator.service.ts]: Hmm yeah it's a bit ugly in either capacity. 

I've re-implemented this pulling the "authenticator skip" logic closer to the pick request method's behavior... It still might not be ideal, but please take a look and let me know what else should be modified.

File: libs/common/src/platform/abstractions/fido2/fido2-authenticator.service.abstraction.ts [hunk 5]
```diff
@@ -145,6 +145,9 @@ export interface Fido2AuthenticatorGetAssertionParams {
 
   // Bypass the UI and assume that the user has already interacted with the authenticator
   assumeUserPresence?: boolean;
+
+  // Used to bypass interaction with the authenticator when the request is initiated by conditional mediation
+  conditionalMediatedAuth?: boolean;
```
Comment (coroiu) [inline libs/common/src/platform/abstractions/fido2/fido2-authenticator.service.abstraction.ts:150]: question: why do we need both of these btw?

File: apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts [hunk 6]
```diff
@@ -241,7 +241,16 @@ export class BrowserFido2UserInterfaceSession implements Fido2UserInterfaceSessi
   async pickCredential({
     cipherIds,
     userVerification,
+    skipUserVerification,
```
Comment (coroiu) [inline apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts:244]: suggestion: if we instead sent in `assumeUserPresence` then we could move all of the "lying about UV" logic into this method

```ts
if (assumeUserPresence && cipherIds.length === 0) {
  return { cipherId: cipherIds[0], userVerified: userVerification };
}
```

**Review by coroiu — COMMENTED** [review]
Yeah, it's meh, 

Edit: I have no idea why I just stopped mid-sentence, but I was referring to the faking, and not the PR! I would never refer to a PR like that 😅

File: apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts [hunk 7]
```diff
@@ -241,7 +241,16 @@ export class BrowserFido2UserInterfaceSession implements Fido2UserInterfaceSessi
...
   }: PickCredentialParams): Promise<{ cipherId: string; userVerified: boolean }> {
+    // NOTE: For now, we are defaulting to a userVerified status of `true` when the request
+    // is for a conditionally mediated authentication. This will allow for mediated conditional
+    // authentication to function without requiring user interaction. This is a product
+    // decision, rather than a decision based on the expected technical specifications.
+    if (skipUserVerification) {
+      return { cipherId: cipherIds[0], userVerified: true };
```
[the remark this answers is not in the document]
Comment (cagonzalezcs) [reply apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts:251]: Done, this change still functions in the way product is expecting.

File: apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts (hunk 6 above)
Comment (cagonzalezcs) [reply apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts]: Yeah, that makes sense. Thank you for the feedback on this. I wasn't really sure about how to properly implement this requirement.

File: libs/common/src/platform/abstractions/fido2/fido2-authenticator.service.abstraction.ts (hunk 5 above)
Comment (cagonzalezcs) [reply libs/common/src/platform/abstractions/fido2/fido2-authenticator.service.abstraction.ts]: Removed, originally I was just trying to ensure that the logic I was triggering was only for conditional mediated auth requests... but I followed your other suggestions, and this is no longer necessary.

File: apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts (hunk 6 above)
Comment (coroiu) [reply apps/browser/src/autofill/fido2/services/browser-fido2-user-interface.service.ts]: Happy to help! Another reason for moving the faking is that the desktop app will be using the same authenticator but does not have to lie. So this way we've isolated the faking to the browser extension!

**Review by coroiu — APPROVED** [review]
Looks good from a platform/passkey perspective! (I mean, except for the UV faking but that's a separate issue!)

**Review by jprusik — APPROVED** [review]
